[Feature] Add S3DIS dataset for 3D object detection

filaPro opened the pull request:
Close #817.

@Wuziyi616 can you have a look about general structure? Then I can also add unit test and documentation. I'm also attaching visualization from `browse_dataset`.

![Screenshot from 2021-08-04 16-30-41](https://user-images.githubusercontent.com/6030962/128191273-a0123c0e-c35f-4417-b07e-c07746146b9d.png)

Wuziyi616 commented on tools/data_converter/s3dis_data_utils.py:
```diff
@@ -99,15 +107,51 @@ def process_single_scene(sample_idx):
         sample_id_list = sample_id_list if sample_id_list is not None \
             else self.sample_id_list
-        with futures.ThreadPoolExecutor(num_workers) as executor:
+        with futures.ThreadPoolExecutor(1) as executor:
```
Please use `num_workers` as thread number (or if there is any reason we can't use multi-threading here?)

Wuziyi616 commented on tools/data_converter/s3dis_data_utils.py:
```diff
@@ -24,13 +24,21 @@ def __init__(self, root_path, split='Area_1'):
+        # Following https://arxiv.org/abs/2006.12356, use 5 furniture classes
+        # for detection: table, chair, sofa, bookcase, board.
+        self.furniture_cat_ids = np.array([7, 8, 9, 10, 11])
+        self.furniture_cat_ids2class = {
```
We may rename `cat_ids` to `seg_cat_ids` and `furniture_cat_ids` to `det_cat_ids` so that their names are more easy to understand. Do you think that's acceptable?

Wuziyi616 commented on tools/data_converter/s3dis_data_utils.py:
```diff
@@ -99,15 +107,51 @@ def process_single_scene(sample_idx):
+        Args:
+            points (np.array): Scene points of shape (n, 6).
+            pts_instance_mask (np.array): Instance labels of shape (n,).
+            pts_semantic_mask (np.array): Semantic labels of shape (n,).
```
The dtype should be `np.ndarray` instead of `np.array`.

Wuziyi616 commented on tools/data_converter/s3dis_data_utils.py:
```diff
@@ -99,15 +107,51 @@ def process_single_scene(sample_idx):
+            pts_semantic_mask (np.array): Semantic labels of shape (n,).
+
+        Returns:
+            infos (dict): Bounding boxes with labels.
```
For the returned values, we often don't write its name here. I think the following docstring is better?
```
Returns:
    dict: A dict containing detection infos with following keys:

        - gt_boxes_upright_depth (np.ndarray): Bounding boxes of shape (n, 6)
        - class (np.ndarray): Box labels of shape (n,)
        - gt_num (int): Number of boxes.
```

Wuziyi616 commented on mmdet3d/datasets/s3dis_dataset.py:
```diff
@@ -294,3 +285,160 @@ def _check_scene_idxs(self, scene_idx, num):
+
+
+class _S3DISDataset(Custom3DDataset):
+    """S3DIS Dataset for Detection Task."""
```
Maybe we can add some docstring like in `_S3DISSegDataset`? For example, the paper link for users' reference. Also the docstring about each input args is required.

Wuziyi616 commented on mmdet3d/datasets/s3dis_dataset.py:
```diff
@@ -294,3 +285,160 @@ def _check_scene_idxs(self, scene_idx, num):
+
+@DATASETS.register_module()
+class S3DISDataset(_S3DISDataset):
+    """S3DIS Dataset for Detection Task."""
```
Also the adding some docstring would be better. E.g. indicating this class is a wrapper over `_S3DISDataset`

Wuziyi616 reviewed (commented):
Good job! @filaPro two questions:
- Similar to ScanNet detection, S3DIS detection also doesn't involve `yaw` prediction right? So the boxes here are all axis-aligned.
- Maybe we can move the two S3DIS detection datasets code to be above the two S3DIS segmentation datasets code? This follows `scannet_dataset.py`

The `browse_dataset` result is great! Only need to add some docstring and minor code structure refinement. Also unit tests are needed.

Wuziyi616 commented:
I let @THU17cyz have a look here because this may involve our progress of refactoring coordinate system.

filaPro commented:
@Wuziyi616,
 - Yes this dataset doesn't have yaw angle.
 - Moved detection dataset to be above the segmentation one.

I also removed _S3DISDataset in favour of ConcatDataset. This also requires a fmall fix in `browse_dataset`. Do you think it is ok now?

THU17cyz commented on mmdet3d/datasets/s3dis_dataset.py:
```diff
@@ -2,14 +2,154 @@
+            index (int): Index of the sample data to get.
+
+        Returns:
+            dict: Data information that will be passed to the data \
```
Same, may remove \

THU17cyz commented:
Great work, many thanks to @filaPro ! Similar to ScanNet det, there is no conflict with our ongoing coordinate system refactor.

Wuziyi616 commented on mmdet3d/datasets/s3dis_dataset.py:
```diff
@@ -2,14 +2,154 @@
 @DATASETS.register_module()
-@SEG_DATASETS.register_module()
+class S3DISDataset(Custom3DDataset):
+    """S3DIS Dataset for Detection Task.
```
Need to add an `r` before `"""`, then the link `GSDN <https://arxiv.org/abs/2006.12356>` will be clickable in the generated docs. See this [line](https://github.com/open-mmlab/mmdetection3d/blob/d3213cd3eb5cd64733844d80ca638ee1f0355bb1/mmdet3d/datasets/s3dis_dataset.py#L14).

Wuziyi616 commented on tools/misc/browse_dataset.py:
```diff
@@ -189,6 +190,10 @@ def main():
+
+    # use only first dataset for ConcatDataset
+    if isinstance(dataset, ConcatDataset):
+        dataset = dataset.datasets[0]
```
Maybe we can move this judgement to between line 64 and line 65 (the line id before modification)? Like:

```
if cfg.data.train['type'] == 'ConcatDataset':
        cfg.data.train = cfg.data.train.datasets[0]
```

Wuziyi616 commented on configs/_base_/datasets/s3dis-3d-5class.py:
```diff
@@ -0,0 +1,113 @@
+        load_dim=6,
+        use_dim=[0, 1, 2, 3, 4, 5]),
+    dict(type='LoadAnnotations3D', with_bbox_3d=True, with_label_3d=True),
+    dict(type='IndoorPointSample', num_points=40000),
```
In PR #667, we unify indoor and outdoor point sampling with a transform `PointSample`. Please kindly change the name.

Wuziyi616 replied:
See this [line](https://github.com/open-mmlab/mmdetection3d/pull/667/files#diff-2d0d31184b759c5edf7a417f47c6994da36a8d640ce4d79a334179f98fa5d003R23)

Wuziyi616 commented on configs/_base_/datasets/s3dis-3d-5class.py:
```diff
@@ -0,0 +1,113 @@
+                sync_2d=False,
+                flip_ratio_bev_horizontal=0.5,
+                flip_ratio_bev_vertical=0.5),
+            dict(type='IndoorPointSample', num_points=40000),
```
Also `PointSample` here

Wuziyi616 reviewed (commented):
Almost ready to merge. Only need to add some unit tests.

filaPro commented:
Don't quite understand lint problems, so revert `yapf: disable` back.

Wuziyi616 commented:
> 
> 
> Don't quite understand lint problems, so revert `yapf: disable` back.

This is because sometimes `isort` and `yapf` will have different behaviors in formatting some codes. You need to disable one of them to pass the linting check.

THU17cyz commented:
Ready to merge after unit tests are added

filaPro commented:
Also updated test data file for S3DIS.

Wuziyi616 reviewed (approved):
Good Job @filaPro! Cheers!
